gossip/emitter/throttling/throttling.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2025 Sonic Operations Ltd
+// This file is part of the Sonic Client
+//
+// Sonic is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Sonic is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with Sonic. If not, see <http://www.gnu.org/licenses/>.
+
+package throttling
+
+import (
+	"github.com/0xsoniclabs/sonic/inter"
+	"github.com/0xsoniclabs/sonic/opera"
+	"github.com/Fantom-foundation/lachesis-base/inter/idx"
+	"github.com/Fantom-foundation/lachesis-base/inter/pos"
+)
+
+//go:generate mockgen -source=throttling.go -destination=throttling_mock.go -package=throttling
+
+// ThrottlingState manages the state for event emission throttling based on
+// the set of dominant validators.
+// This type contains the state needed to decide whether to skip event emission
+// for a given validator, based on its stake and the stake distribution among
+// all validators, and properties of the events to be emitted.
+type ThrottlingState struct {
+	// throttler configuration parameters
+	thisValidatorID                     idx.ValidatorID
+	dominatorsThreshold                 float64
+	maxSkippedEventsWithSameFrameNumber uint
+
+	// means to access the world state
+	world WorldReader
+
+	// accumulated state
+	lastUsedFrame             idx.Frame
+	lastEmissionBlockNumber   idx.Block
+	currentSkippedEventsCount uint
+}
+
+type WorldReader interface {
+	GetRules() opera.Rules
+	GetLatestBlockIndex() idx.Block
+	GetEpochValidators() (*pos.Validators, idx.Epoch)
+}
+
+// NewThrottlingState creates a new ThrottlingState for a given validator ID
+// and dominating percentile threshold.
+//
+// dominatingPercentile parameter specifies the fraction of total stake
```

Comment:
nit: about terminology - the [percentile](https://en.wikipedia.org/wiki/Percentile) is the actual value, not the percentage. In this context, it would be more accurate to refer to the `dominatingPercentile` value as a `minimumActiveValidatorFraction` or `minimumLiveValidatorFraction`.